api/envoy/service/discovery/v3/discovery.proto

```diff
@@ -110,6 +111,15 @@ message DiscoveryResponse {
   // [#not-implemented-hide:]
   // The control plane instance that sent the response.
   config.core.v3.ControlPlane control_plane = 6;
+
+  // Time-to-live value for the response. A timer is started that is reset each time the
+  // response is received with a new TTL. If the response is received with no TTL set, the timer is removed for the
+  // response. Upon expiration of the timer, all the resources contained in the response will be removed.
+  //
+  // The TTL feature is meant to support configurations that should be removed in the event of a management server
+  // failure. For example, the feature may be used for fault injection testing where the fault injection should be
+  // terminated in the event that Envoy loses contact with the management server.
+  google.protobuf.Duration ttl = 7;
```

Comment:
As we discussed previously in #10898 (see https://github.com/envoyproxy/envoy/pull/10898#discussion_r414129591), having the TTL in the `DiscoveryResponse` for the SotW protocol variant yields extremely unclear and confusing semantics.  In that PR, we agreed that the way we'd do this in the SotW protocol variants was to have the `Any` field contain a `Resource` object.  That way the TTL semantics are the same in the SotW and incremental protocol variants.

I think we should stick to the approach that we agreed upon there.  The semantics are going to be way too unclear otherwise.

Reply:
Thanks for linking that, I hadn't seen that discussion. 

I also wasn't aware of the fact that we can end up with multiple responses making up the full state for SotW, that definitely complicates the implementation. This PR simply clears out all the resources for the type url once a TTL is hit, but that doesn't sound correct anymore.

Adding per resource TTLs effectively means that we need to track per resource state for SotW when using TTLs, which definitely makes this more complicated. Curious what @htuch @mattklein123 thinks about this

Reply:
Sorry this has been in flight for so long that I can't remember the details anymore. Conceptually it seems like the TTL can be for the entire SoTW response. Can you summarize why that was rejected?

Reply:
The details are in the thread at https://github.com/envoyproxy/envoy/pull/10898#discussion_r414129591.

Basically, the problem is that the SotW response does not necessarily have to include every single resource of that type.  That's true for Listener and Cluster resources as special cases, but a SotW response for any other resource type can include only those resources that have changed.   For details, see https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#grouping-resources-into-responses.

Here's the concrete example I cited in the other thread:

> For example, for EDS, let's say that at time t=0, the client receives a response containing resource A that does not set the TTL, and then at time t=1m, it receives another response containing resource B that does set the TTL to 5m. Does that mean that the TTL in the second response would be used for both A and B? I'm guessing both. Would the expiration time for A be t=6m, or would it be t=5m (retroactive from when A was received)?

We can certainly define what the semantic should be in this case, but I think any semantic we pick will be fairly counter-intuitive and confusing.  And I think the fact that SotW and incremental APIs will have different semantics from each other will also be a source of confusion.

If we're going to accept per-resource timeouts in the incremental API, I think we should just do the same thing in SotW.  That will provide more consistency and less confusion for everyone.

Reply:
Now I'm really confused. Is ^ actually true for SoTW? In the case of EDS, each EDS subscription would be independent and is then SoTW. Or is this a complexity of ADS? Given multiple independent EDS subscriptions wouldn't the TTL just apply independently to each subscription at the SoTW level?

Reply:
I believe you had exactly the same reaction in the thread I mentioned. :)  I would argue that the fact that we API shepherds can't keep this straight in our heads is a good argument for working to eliminate the SotW protocol variants, since the incremental variants don't suffer from these kinds of confusing special cases.

In answer to your question, @htuch can chime in here when he gets back to the office (he's out today), but my understanding is that what I said is true for all xDS resource types except for Listener and Cluster.  I believe that for the non-ADS EDS case, Envoy's current implementation does happen to create a separate stream for each resource, which means that this issue won't actually show up there, but I think that's just an side effect of that implementation detail; from the protocol perspective, there's no reason it could not request all resources on one EDS stream, in which case it would behave as I described.  In the ADS case, I believe that Envoy does get all resources on the same stream, and it does behave as I described.  So this may be an ADS-specific problem in practice, but I think that's just a coincidence, not a requirement of the protocol.

Reply:
> I would argue that the fact that we API shepherds can't keep this straight in our heads is a good argument for working to eliminate the SotW protocol variants, since the incremental variants don't suffer from these kinds of confusing special cases.

That may be true, but that's aspirational. The will be around for years so we need to deal with it.

OK let's wait for @htuch to come back and maybe let's have a short meeting to hash this out in a more high bandwidth environment. I agree this is incredibly confusing.

Reply:
> That may be true, but that's aspirational. The will be around for years so we need to deal with it.

Yes, absolutely.  I'm not arguing for any sort of overnight change, just that it's something we should be trying to move towards in an incremental (no pun intended :) ) fashion.  I think that the "xds transport next steps" design is one nice step toward that goal.  Adding key frame support to the incremental protocol variants would also be really helpful IMHO.

In any case, I certainly agree that this is a broader discussion; I mentioned here only to point out that it's something we should be thinking about.

Reply:
We have two classes of xDS SotW APIs today:
1. Collection (wildcard): LDS/CDS
2. Resource subscription: RDS/EDS/SDS

Class 2 create subscriptions, whether on individual streams or ADS, and can receive resource updates, but never delete resources. They rely on something in Class 1 to cause an unsubscription. So, we can easily set TTLs on class 1 and have it apply to all resources of the type. For Class 2 we really have each DiscoveryResponse implying a TTL for 1 or more resources.

I agree it's kind of terrible, but it is what it is. I think what we could is have folks include TTLs in Resources as discussed previously (I had also semi-forgotten this conclusion), and Envoy can optimize in deduping timers when it has a bunch of resources with the same deadline.

Reply:
Seems to me like adding a `Resource` wrapper to SotW would introduce per resource versioning to the API, which would also be confusing? Given that the version field would be completely ignored.

Part of me feels like it would be okay to let SotW not be concerned with individual resources and have the TTL always apply to all resources for current stream, even if the responses might not necessarily contain all the resources. While each response might only contain a subset of the resources, its all coming from the server so it would be up to the server to sequence things in the right way to get TTLs working.

I get that it's confusing to have Delta and SotW work differently, but it also seems confusing to partially backport per resource support to SotW

Reply:
I think it's fine to ignore the per-resource version field in the SotW protocol variants.  After all, the incremental protocol variants still provide a top-level resource type instance version, but it doesn't get used for anything there, because the ACK/NACK is based on the individual resource version.  So this just makes things symmetrical: both protocol variants provide both top-level and per-resource versions, but they differ based on which one they use for ACK/NACK.

I think that it's actually a good thing to start to support `Resource` in the SotW protocol variants, since that's a small step toward eliminating a potential difference between SotW and incremental.  And if we control it with a client capability, as we originally agreed, then it's a fully backward-compatible change.  That seems a lot better to me than having TTLs work differently in SotW and incremental, because it moves the two closer together instead of adding yet another difference between them.

If we are okay with applying TTLs for the whole resource type in SotW, we could also decide to do the same thing for incremental.  That would be another way of keeping things consistent between them.

I think perhaps one of the underlying questions here is, do we consider the TTL to be part of the transport protocol or part of the data model?  If we consider it part of the transport protocol, then it makes sense to set it globally for a whole resource type.  But if we consider it part of the data model, then we should make it an attribute of the individual resources.

Reply:
TTL is clearly transport protocol AFAICT.

Reply:
Okay, then why don't we say that this is top-level value for both SotW and incremental?

Reply:
Seems like we're going back a forth a bit, so I'll try to summarize the issue and open questions in a short doc and try to schedule a meeting. Hopefully that will make it easier for us to come to agreement here.

Reply:
Thanks @snowp. I think TTLs are a power feature for API scalability, as well as the reliability concerns, so really appreciate you driving this, look forward to a plan-of-record. Happy to dive into implementation details once design is settled.

Reply:
Sharing the doc here for anyone else following along: https://docs.google.com/document/d/1SSF9xMQpxJE_22khEz7CFbc0tddjxLjeuVEx05pg2uU/edit#